### ISSUE TITLE
Asset missing from graph

### ISSUE THREAD

**Issue by topherbuckley** [eprbell/dali-rp2#156]
Getting the following error when running via the bitbank rest api plugin. This is using the tip of my makerfees branch with PR [here](https://github.com/eprbell/dali-rp2/pull/155) but hoping this has nothing to do with it and this error was just hidden under the other errors I was having before. Does this mean XYM is not supported for price lookup? I'll keep reading the code and see if I can make sense of it in the meantime. 

```
ERROR: Fatal exception occurred:
Traceback (most recent call last):
  File "/media/christopher/HDD/git/dali-rp2/src/dali/dali_main.py", line 181, in _dali_main_internal
    resolved_transactions: List[AbstractTransaction] = resolve_transactions(transactions, dali_configuration, args.read_spot_price_from_web)
  File "/media/christopher/HDD/git/dali-rp2/src/dali/transaction_resolver.py", line 263, in resolve_transactions
    transaction = _update_spot_price_from_web(transaction, global_configuration)
  File "/media/christopher/HDD/git/dali-rp2/src/dali/transaction_resolver.py", line 136, in _update_spot_price_from_web
    conversion: RateAndPairConverter = _get_pair_conversion_rate(
  File "/media/christopher/HDD/git/dali-rp2/src/dali/transaction_resolver.py", line 108, in _get_pair_conversion_rate
    rate = cast(AbstractPairConverterPlugin, pair_converter).get_conversion_rate(timestamp, from_asset, to_asset, exchange)
  File "/media/christopher/HDD/git/dali-rp2/src/dali/abstract_pair_converter_plugin.py", line 178, in get_conversion_rate
    historical_bar = self.get_historic_bar_from_native_source(timestamp, from_asset, to_asset, exchange)
  File "/media/christopher/HDD/git/dali-rp2/src/dali/plugin/pair_converter/ccxt.py", line 241, in get_historic_bar_from_native_source
    raise RP2RuntimeError(f"The asset {from_asset} or {to_asset} is missing from graph")
rp2.rp2_error.RP2RuntimeError: The asset XYM or USD is missing from graph
```

**Comment by topherbuckley** [comment]
Just ran this in a debugger and found that the None variable was `from_asset_vertex` which in this case is defined as:

`from_asset_vertex: Optional[Vertex[str]] = current_graph.get_vertex(from_asset)`

with `from_asset` being XYM. From a conceptual standpoint, can someone tell me what a "graph" and "vertex" are supposed to represent in this context? I'm having trouble debugging as I'm not sure what the `current_graph` is supposed to look like. The definition being:

`current_graph = self.__exchange_graphs[exchange]`, and `exchange` being 'Kraken' at the time when this runtime error fires. I don't see the definition of `__exchange_graphs` anywhere, so if someone can clue me into how this is populated, I may be more successful and finding the cause of this.

**Comment by macanudo527** [comment]
Basically, you are trying to price an asset that isn't available on the default exchange - Kraken. You should be able to resolve this by manually adding the alternative market.

To do that, you need to add a market to an exchange and then add a market. Add the market to the `_ALT_MARKET_EXCHANGES_DICT` on [line #99](https://github.com/eprbell/dali-rp2/blob/483baa3e89a3b18e0f853a0c504d4de3dc28c5ae/src/dali/plugin/pair_converter/ccxt.py#L99) by adding the following line (in alphabetical order please):
```
    "XYMUSDT": _GATE,
```
And then add that market to the `_ALT_MARKET_BY_BASE_DICT` on [line #101](https://github.com/eprbell/dali-rp2/blob/483baa3e89a3b18e0f853a0c504d4de3dc28c5ae/src/dali/plugin/pair_converter/ccxt.py#L101):
```
    "XYM": "USDT",
```

That should resolve the issue. It will use the Gate.io market for pricing, [which seems to have the most volume](https://coinmarketcap.com/currencies/symbol/) and is an exchange we support currently. We don't have support for Zaif or Bithumb.

Basically, the graph is what is used by the Dijkstra search to find the shortest route between two assets so that the CCXT pair converter can price it. The vertexes are markets that connect the nodes (crypto or fiat assets). Right now some of that graph has to be manually built and optimized. I apologize for the mess, I'm actually knee-deep in a process that should clean this up and make this process a lot easier to dynamically create routing instead of manually adding these.

**Comment by topherbuckley** [comment]
@macanudo527 thanks a lot for the help. I'm guessing you don't want me to add a PR for this right? Just a one-off fix for my case, and whatever you're working on would override this? 

>Basically, the graph is what is used by the Dijkstra search to find the shortest route between two assets so that the CCXT pair converter can price it. The vertexes are markets that connect the nodes (crypto or fiat assets). 

Neat! Got it, thanks.

**Comment by macanudo527** [comment]
Well, if you have the time to submit one, it would be nice of you to add a PR. Realistically, it might take me a month or two to get there and someone might find it useful in the meantime. Anyway, it's only two lines so it's easy to submit/approve/revoke.

**Comment by topherbuckley** [comment]
No problem! I'll wait until I finish up a few more adds there, as it looks like there were other unsupported ones in there as well.

**Comment by topherbuckley** [comment]
Oh, as an aside, is there a good existing unit test to test this? Running the whole rest api import from bitbank takes a while, and I might have a few more of these to handle.

**Comment by macanudo527** [comment]
Since it depends on real-world data it is hard to test for. Theoretically, Kraken (or Binance if you lock the CCXT plugin to that) could add one of these assets and then it would be supported automatically. It's only when a user has an asset that isn't on Kraken (since it is the default) that this pops up.

Once graph optimization gets in place we can scan more major exchanges (Gate.io, Coinbase, etc...) and add these assets automatically. That should support everything, but dead or meme coins.

**Comment by topherbuckley** [comment]
Just curious, and I'm not sure exactly what is involved in "graph optimization", but what about just implementing what you manually did for XYM? i.e. use the the coinmarketcap api to check which of the currently supported exchanges on dali is the highest ranking by market cap. I'm sure you have a much better idea of what is best here, so please feel free to ignore, just curious.

**Comment by macanudo527** [comment]
Well, the coinmarketcap api doesn't have historical data for free. Even the first paid tier only offers 1 year of historical data. We would probably need the last 2-4 years to be safe and we can get those for free from major exchanges.

The idea is for things to work out of the box and not require users to sign up for too many API keys. Although, some people have talked about using coinmarketcap or coingecko api as a backup for pricing. I'm not sure how useful it would be if people are just using the free tier, though.

**Comment by topherbuckley** [comment]
I see, yeah that makes sense.
